Private Static Fields Features: Stage 3

rricard opened the pull request:
| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | #8052 (except accessors)
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Add private static fields support
| Tests Added + Pass?      | Tests Updated
| Documentation PR         | <!-- If so, add `[skip ci]` to your commit message to skip CI -->
| Any Dependency Changes?  |
| License                  | MIT
| Sponsor                 | @bloomberg

- No changes in parser
- Added a new helper to check private static field access provenance: `classStaticPrivateFieldBase`
- While visiting a class:
    1. Check for duplicate private static declarations
    2. For every private static declaration add it to a container object in the classe's scope and remove it from the class itself
        ```js
        class C {
          static #foo = "bar";
          /*...*/
        }
        // will become (in loose mode):
        class C {
          /*...*/
        }
        var _CStatics = {};
        _CStatics._foo = "bar";
        ```
    3. For every private static declaration, start visiting the class to explore each member method in order to wrap the access in the helper:
        ```js
        myMethod() {
          C.#foo = this.#foo + "baz";
        }
        // Becomes
        myMethod() {
          babelHelper.classStaticPrivateFieldBase(C, C, _CStatics)._foo = 
            babelHelper.classStaticPrivateFieldBase(this, C, _CStatics)._foo + "baz";
        }
        ```

rricard commented:
To do list:

- [ ] Review all current test outputs one by one, make sure we run `exec.js` versions when possible
- [ ] Private static setters/getters
- [ ] Account for a spec version

rricard commented:
We only handled private static fields for now, @tim-mc is going to help bring private static methods now. From there we'll be able to think about accessors...

rricard commented:
It feels like this PR is ready for review.

I also would like some help for making that Travis build pass...

hzoo commented on packages/babel-plugin-proposal-class-properties/test/fixtures/private-loose/static-undefined/exec.js:
```diff
@@ -0,0 +1,15 @@
+
+expect("bar" in Foo).toBe(false)
+expect(Foo.test()).toBe(undefined)
+expect(Foo.test()).toBe(undefined)
```
Some of these duplicated tests are weird to me `expect(Foo.test()).toBe(undefined)` - I know they were just copied over but is it supposed to be testing the static and instance methods or just remove it?

nicolo-ribaudo replied:
Maybe it should be `expect(() => Foo.test()).not.toThrow()`?

rricard replied:
@hzoo, you're right, I'll go ahead and do a second review pass on those tests (was only checking for the output I was seeing). I'm also trying to run test-262 right now against it (having `npm link` issues) and I might add some relevant tests then

jridgewell commented on packages/babel-helpers/src/helpers.js:
```diff
@@ -1011,3 +1011,12 @@ helpers.classPrivateFieldSet = () => template.program.ast`
+
+helpers.classStaticPrivateFieldBase = () => template.program.ast`
+  export default function _classStaticPrivateFieldBase(receiver, classConstructor, privateClass) {
+    if (receiver !== classConstructor && receiver.constructor !== classConstructor) {
```
The `receiver.constructor !== classConstructor` shouldn't be necessary.

rricard replied:
Unfortunately, this seems necessary to be spec compliant. We should discuss that on slack.

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -285,14 +377,18 @@ export default declare((api, options) => {
-              throw path.buildCodeFrameError(
-                "Static class fields are not spec'ed yet.",
-              );
+              if (privateStaticNames.has(name)) {
```
Private names are shared between instance and static, ie, the following is an error:

```js
class Example {
  static #x;
  #x;
}
```

rricard replied:
Wasn't sure about this one, I'll make the change.

jridgewell commented on packages/babel-plugin-proposal-class-properties/test/fixtures/private-loose/native-classes/output.js:
```diff
@@ -1,18 +1,37 @@
+var Foo =
+/*#__PURE__*/
+function () {
+  "use strict";
```
This is my mistake. This test case was meant to output `class` expressions, not compile classes to functions.

rricard replied:
That makes more sense indeed, I'll redo the whole test dir with that then!

rricard replied:
@jridgewell what should I do to convert tests to output class expressions?

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -83,6 +83,18 @@ export default declare((api, options) => {
   };
 
+  // Traverses the class scope, handling private static name references.
+  const staticPrivatePropertyVisitor = {
```
We should be able to reuse `privateNameVisitor`. This is necessary because we have to use its inner traverser, too.

rricard replied:
Ok, I kept them separate to make sure I don't mess with your logic but indeed, it's better to change just a few things in your logic than to branch out an entirely new one!

jridgewell commented on packages/babel-helpers/src/helpers.js:
```diff
@@ -1011,3 +1011,12 @@ helpers.classPrivateFieldSet = () => template.program.ast`
 `;
+
+helpers.classStaticPrivateFieldBase = () => template.program.ast`
+  export default function _classStaticPrivateFieldBase(receiver, classConstructor, privateClass) {
```
So there are a few issues here, and we can choose either of two ways to handle it:

- We can remove this `privateClass` holder object for `loose` mode.
- We can handle `get`, `set`, and `call` for both `loose` and `spec` mode.

The first option would allow us to reuse the same code we already have for instance private fields.

If we continue to use a holder object, we have to handle `get`, `set`, and `call` operations on the holder, and use the `receiver` as the calling context for all those cases. This complicates the runtime for `loose` mode. But, we're gonna need to do that anyways for spec mode.

rricard replied:
This is super helpful, I had a hard time understanding the line between loose and spec.

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -157,6 +169,23 @@ export default declare((api, options) => {
   };
 
+  const staticPrivatePropertyHandler = {
+    handle(member) {
```
So this doesn't handle `get`, `set`, and `call` operations. Given that all this is stored on a holder object, anything like `ClassName.#x()` will really have to be transformed to `base(ClassName, ClassName, holder)['x'].call(ClassName)`. This currently just outputs `base(ClassName, ClassName, holder)['x']()`.

rricard replied:
I think I just missed something about how the handlers are working and that made it way clearer. Thanks!

rricard commented:
Thanks @jridgewell, I'm sorry I have a lot of work on the side of this but I'll try to get back to it this week.

rricard commented:
Todo:

- [ ] Rework the helper(s)
- [ ] Redo the tests
- [ ] get/set/call handler
    - [ ] spec
    - [ ] loose

rricard commented:
Hi @jridgewell, I could use some help changing the tests to not transform the classes. Otherwise, I think I'm almost done.

jridgewell commented on packages/babel-helpers/src/helpers.js:
```diff
@@ -1029,3 +1029,34 @@ helpers.classPrivateFieldSet = () => template.program.ast`
+
+helpers.classStaticPrivateFieldLooseBase = () => template.program.ast`
+  export default function _classStaticPrivateFieldLooseBase(receiver, classConstructor) {
+    if (receiver !== classConstructor && receiver.constructor !== classConstructor) {
```
What's the `receiver.constructor !== classConstructor` for?

rricard replied:
Rereading the whole thing, this logic is busted. That's why I want to make sure the tests work as well

rricard replied:
That's for when the static field gets called from `this`. In that case I'll probably have an instance of the class instead of the constructor itself. I do think that's what we expect from the spec but it's been a while since the last time I dived in the proposal.

jridgewell replied:
It has to throw a type error if it's called on an instance of the class. This check should just be the `receiver !== classConstructor`.

rricard replied:
I'm not sure, I think that if the instance is of the class type, access should work through `this`. I can check that again. However, in the meantime, we can restrict the access so it does not work on the instance while we figure that out.

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -245,6 +299,77 @@ export default declare((api, options) => {
+      // Create a private static "host" object if it does not exist
+      privateClassId = path.scope.generateUidIdentifier(ref.name + "Statics");
+      staticNodesToAdd.push(
+        template.statement`const PRIVATE_CLASS_ID = {};`({
```
Nit: Should probably use a prototype-less object here.

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -278,19 +407,13 @@ export default declare((api, options) => {
-            }
             if (privateNames.has(name)) {
-              throw path.buildCodeFrameError("Duplicate private field");
+              throw path.buildCodeFrameError("Duplicate static private field");
```
Let's leave this as the old message.

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -353,11 +476,27 @@ export default declare((api, options) => {
+                state,
+                privateClassId,
+              );
+              staticNodesToAdd.forEach(node => staticNodes.push(node));
```
Can just `staticNodes.push(...staticNodesToAdd)`

rricard commented:
@jridgewell don't worry, we're in no rush. I'll just need some help setting up the tests correctly.

jridgewell commented:
You can probably duplicate some of the tests from the instance private fields, we just need to change the `input.js` code to use statics.

rricard commented:
I was trying to prevent the class transform itself but I managed to get that right in the plane

rricard commented:
It should be good for a second review

jridgewell commented on packages/babel-plugin-proposal-class-properties/src/index.js:
```diff
@@ -245,6 +299,77 @@ export default declare((api, options) => {
+    const { scope, parentPath } = path;
+    const { key, value } = path.node;
+    const { name } = key.id;
+    const privateId = scope.generateUidIdentifier(name);
```
Nit: So we actually don't need to generate a unique id here, we can just use `name`.

rricard commented:
Loose tests have been removed while I changed my test methodology, they're back now. I also removed the privateId for spec mode but not loose mode (where the unique id prevents test breakage in loose mode)

rricard commented:
Rebased the whole thing

rricard commented:
@jridgewell I think this covers all of the requested changes. Thanks!

jridgewell commented on packages/babel-plugin-proposal-class-properties/test/fixtures/private-loose/native-classes/output.js:
```diff
@@ -0,0 +1,27 @@
+
+}
+
+Foo._foo = "foo";
```
My last nit: This should be using `Object.defineProperty` with `{ enumerable: false, configurable: false }`. We can do this in a follow PR if you'd like.

rricard replied:
That is a good point, I'll try to put that here before we merge

rricard commented:
@jridgewell I made the change, would you like to have it in a separate helper though?

jridgewell commented:
Eh, let's just go for it.

nicolo-ribaudo commented:
I'm sorry, I forgot to review this PR :sweat_smile:
I'll open a few PR with some small tweaks.
